File: web/src/lib/watches.ts
import { sql } from "@/lib/db";

const SOURCES = ["company_page", "radar_onboarding", "resume"] as const;
export type WatchSource = (typeof SOURCES)[number];

export function isWatchSource(value: string): value is WatchSource {
  return (SOURCES as readonly string[]).includes(value);
}

export async function followCompany(
  userId: string,
  companyId: string,
  source: WatchSource,
): Promise<void> {
  if (!sql) throw new Error("database unconfigured");
  await sql`
    insert into watches (user_id, company_id, status, source)
    values (${userId}, ${companyId}::uuid, 'active', ${source})
    on conflict (user_id, company_id) do update
      set status = 'active',
          source = excluded.source,
          updated_at = now()
  `;
  await sql`
    insert into product_events (user_id, name, company_id)
    values (${userId}, 'follow', ${companyId}::uuid)
  `;
}

export async function unfollowCompany(userId: string, companyId: string): Promise<void> {
  if (!sql) throw new Error("database unconfigured");
  await sql`
    update watches
    set status = 'unwatched', updated_at = now()
    where user_id = ${userId} and company_id = ${companyId}::uuid
  `;
  await sql`
    insert into product_events (user_id, name, company_id)
    values (${userId}, 'unfollow', ${companyId}::uuid)
  `;
}

export async function companyIdForSlug(slug: string): Promise<string | null> {
  if (!sql) return null;
  const rows = await sql<{ id: string }[]>`
    select id from companies where slug = ${slug} limit 1
  `;
  return rows[0]?.id ?? null;
}

export async function isWatching(userId: string, companyId: string): Promise<boolean> {
  if (!sql) return false;
  const rows = await sql<{ status: string }[]>`
    select status from watches
    where user_id = ${userId} and company_id = ${companyId}::uuid
    limit 1
  `;
  return rows[0]?.status === "active";
}

export type RadarEvent = {
  id: string;
  headline: string;
  summary: string;
  materiality: string;
  published_at: string;
  slug: string;
  name: string;
};

export type WatchedCompany = {
  slug: string;
  name: string;
  category: string;
};

export async function listRadarEvents(userId: string, limit = 40): Promise<RadarEvent[]> {
  if (!sql) return [];
  return sql<RadarEvent[]>`
    select
      e.id,
      e.headline,
      e.summary,
      e.materiality,
      e.published_at,
      c.slug,
      c.name
    from change_events e
    join companies c on c.id = e.company_id
    join watches w on w.company_id = c.id
    where w.user_id = ${userId}
      and w.status = 'active'
      and e.publication_state = 'published'
      and e.materiality = 'material'
    order by e.published_at desc, e.id desc
    limit ${limit}
  `;
}

export async function listWatchedCompanies(userId: string): Promise<WatchedCompany[]> {
  if (!sql) return [];
  return sql<WatchedCompany[]>`
    select c.slug, c.name, c.category
    from watches w
    join companies c on c.id = w.company_id
    where w.user_id = ${userId} and w.status = 'active'
    order by c.name
  `;
}

export async function recordRadarView(userId: string): Promise<void> {
  if (!sql) return;
  await sql`
    insert into product_events (user_id, name)
    values (${userId}, 'radar_view')
  `;
}

export async function catalogSuggestions(): Promise<WatchedCompany[]> {
  if (!sql) return [];
  return sql<WatchedCompany[]>`
    select slug, name, category
    from companies
    order by category, name
    limit 24
  `;
}

File: worker/privacyradar/watches.py
"""Watchlist mutations. User identity is always a session-derived argument."""

from __future__ import annotations

from typing import Any

import psycopg

ALLOWED_SOURCES = frozenset({"company_page", "radar_onboarding", "resume"})


class WatchError(ValueError):
    """Caller supplied an untrusted user identifier or invalid watch input."""


def session_user_id(*, session_user_id: str | None, body: dict[str, Any] | None = None) -> str:
    payload = body or {}
    if "userId" in payload or "user_id" in payload:
        raise WatchError("user id must come from session")
    if not session_user_id:
        raise WatchError("missing session")
    return session_user_id


def follow(
    conn: psycopg.Connection[dict[str, Any]],
    *,
    user_id: str,
    company_id: str,
    source: str,
) -> None:
    if source not in ALLOWED_SOURCES:
        raise WatchError("invalid source")
    conn.execute(
        """
        insert into watches (user_id, company_id, status, source)
        values (%s, %s, 'active', %s)
        on conflict (user_id, company_id) do update
          set status = 'active',
              source = excluded.source,
              updated_at = now()
        """,
        (user_id, company_id, source),
    )
    conn.execute(
        """
        insert into product_events (user_id, name, company_id)
        values (%s, 'follow', %s)
        """,
        (user_id, company_id),
    )


def unfollow(
    conn: psycopg.Connection[dict[str, Any]],
    *,
    user_id: str,
    company_id: str,
) -> None:
    conn.execute(
        """
        update watches
        set status = 'unwatched', updated_at = now()
        where user_id = %s and company_id = %s
        """,
        (user_id, company_id),
    )
    conn.execute(
        """
        insert into product_events (user_id, name, company_id)
        values (%s, 'unfollow', %s)
        """,
        (user_id, company_id),
    )


def list_radar_events(
    conn: psycopg.Connection[dict[str, Any]],
    *,
    user_id: str,
    limit: int = 40,
) -> list[dict[str, Any]]:
    rows = conn.execute(
        """
        select
          e.id,
          e.headline,
          e.summary,
          e.materiality,
          e.published_at,
          e.publication_state,
          c.slug,
          c.name
        from change_events e
        join companies c on c.id = e.company_id
        join watches w on w.company_id = c.id
        where w.user_id = %s
          and w.status = 'active'
          and e.publication_state = 'published'
          and e.materiality = 'material'
        order by e.published_at desc, e.id desc
        limit %s
        """,
        (user_id, limit),
    ).fetchall()
    return list(rows)


def list_active_watches(
    conn: psycopg.Connection[dict[str, Any]],
    *,
    user_id: str,
) -> list[dict[str, Any]]:
    rows = conn.execute(
        """
        select w.company_id, w.status, w.source, c.slug, c.name, c.category
        from watches w
        join companies c on c.id = w.company_id
        where w.user_id = %s and w.status = 'active'
        order by c.name
        """,
        (user_id,),
    ).fetchall()
    return list(rows)
